logging: support custom timestamp formatting

heinwessels opened the pull request:
This feature adds support for using a custom function to format the timestamp used for all logging backends. When the kconfig option `LOG_OUTPUT_FORMAT_CUSTOM_TIMESTAMP` is set the custom formatting function will always be used.

I'm not sure if this is the right/good approach.If I get a thumbs up I will add unit tests.

---

The reason for this pull request is the need for printing the timestamp in UTC format `hh:mm::ss.ms`. It seems clunky adding adding another `if(IS_ENABLED(CONFIG_LOG_OUTPUT_FORMAT_UTC_TIMESTAMP))`, so I thought it would be better adding functionality to add _any_ format.

And example formatting function I wrote (in C++):
```C++
        static int _printTimeStampUTC(const struct log_output* output,
                                      const log_timestamp_t timestamp,
                                      const log_timestamp_printer_t printer) {
            char time_str[sizeof("00:00:00")];

            time_t time = timestamp / 1000;
            struct tm* tm = gmtime(&time);
            strftime(time_str, sizeof(time_str), "%T", tm);

            uint16_t milliseconds = timestamp % 1000;
            return printer(output, "%s.%03d ", time_str, milliseconds);
        }
```

Which outputs something like:
```
08:22:51.002 <inf> my_app: This is the message that I logged
```

heinwessels committed:
logging: support custom timestamp formatting

This feature adds support for using a custom function to format
the timestamp used for all logging backends. When the kconfig
option LOG_OUTPUT_FORMAT_CUSTOM_TIMESTAMP is set the custom
formatting function will always be used.

Signed-off-by: Hein Wessels <heinwessels93@gmail.com>

nordicjm commented:
@nordic-krch can you review this please?

xudongzheng commented:
This should also resolve #47917.

nordic-krch commented on subsys/logging/log_output_custom.c:
```diff
@@ -20,3 +23,25 @@ void log_custom_output_msg_set(log_format_func_t format)
 }
+#endif /* defined(CONFIG_LOG_CUSTOM_FORMAT_SUPPORT) */
+
+#if defined(CONFIG_LOG_OUTPUT_FORMAT_CUSTOM_TIMESTAMP)
```
can you check if removing those ifdefs impacts the code size? i think that there might be a linker garbage collection and function is removed when not used.

heinwessels replied:
You are correct. I removed the `#if`s and the code size doesn't change.

nordic-krch reviewed (changes requested):
Looks good. Please add a test for that feature.

heinwessels committed:
tests: subsys: logging: timestamp: add custom timestamp tests

Adds tests for the custom timestamp functionality. It also
includes a basic test for the default behaviour.

Signed-off-by: Hein Wessels <heinwessels93@gmail.com>

heinwessels commented:
I added a unit test. I created a new set of tests called `tests/subsys/logging/log_timestamp` because adding the `LOG_OUTPUT_FORMAT_CUSTOM_TIMESTAMP` is quite intrusive on other existing tests. This also adds a nice location for any future tests regarding the timestamp itself.

heinwessels commented:
The current CI failure seems to be the same as reported here: https://github.com/zephyrproject-rtos/zephyr/issues/56349

heinwessels commented:
Now I have a different unrelated error:
```
INFO    - 607/607 qemu_riscv32_smp          tests/subsys/mgmt/mcumgr/cb_notifications/cb.notifications  FAILED Failed (qemu 2.055s)
```
```
START - test_notifications_enabled

    Assertion failed at WEST_TOPDIR/zephyr/tests/subsys/mgmt/mcumgr/cb_notifications/src/main.c:171: callback_enabled_test_notifications_enabled: (cmd_done_got is false)
Expected done command callback

 FAIL - test_notifications_enabled in 0.005 seconds
```
Tried reproducing it locally, but it passes fine.
```
scripts/twister -T tests/subsys/mgmt/mcumgr/cb_notifications/ -p qemu_riscv32_smp
```

heinwessels commented:
Attempted a rebase to `main`.

heinwessels commented:
@nordic-krch CI seems to be all green now, which means it's ready for a re-review :)

nordic-krch commented on tests/subsys/logging/log_timestamp/src/log_timestamp_test.c:
```diff
@@ -0,0 +1,102 @@
+
+ZTEST(test_timestamp, test_custom_timestamp)
+{
+	log_custom_timestamp_set(custom_timestamp);
```
to avoid code repetition you can do following and have one test case:
```
if (IS_ENABLED(CONFIG_LOG_OUTPUT_FORMAT_CUSTOM_TIMESTAMP)) {
   log_custom_timestamp_set(custom_timestamp);
}
static const char *exp_str = (IS_ENABLED(CONFIG_LOG_OUTPUT_FORMAT_CUSTOM_TIMESTAMP) ?
		"custom-timestamp: " DNAME "/" SNAME ": " TEST_STR "\r\n" :
                "[00000000] " DNAME "/" SNAME ": " TEST_STR "\r\n";
```

nordic-krch commented on tests/subsys/logging/log_timestamp/testcase.yaml:
```diff
@@ -0,0 +1,9 @@
+    - native_posix
+
+tests:
+  logging.log_output_custom_timestamp:
```
since test could also be run in configuration without custom timestamp function add:
```
logging.log_output_default_timestamp:
    tags: log_output logging timestamp
```

heinwessels replied:
Added.
